common/scala/src/main/scala/whisk/core/mesos/MesosContainerFactory.scala

```diff
@@ -0,0 +1,163 @@
+                            taskIdGenerator: () => String = MesosContainerFactory.taskIdGenerator)
+    extends ContainerFactory {
+
+  val subscribeTimeout = 30.seconds
+
+  //init mesos framework:
+  implicit val as: ActorSystem = actorSystem
+  implicit val ec: ExecutionContext = actorSystem.dispatcher
+
+  //mesos master url to subscribe the framework to
+  val mesosMaster = mesosConfig.masterUrl
+  //public mesos url where developers can browse logs (till there is way to delegate log retrieval to an external system)
+  val mesosMasterPublic = mesosConfig.masterPublicUrl
+
+  var isSubscribed = false;
+
+  val mesosClientActor = clientFactory(as, mesosConfig)
+
+  //periodically retry if subscribing did not suceed
+  as.scheduler.schedule(0.seconds, (subscribeTimeout.toSeconds + 10).seconds) {
+    if (!isSubscribed) {
+      subscribe()
+    }
+  }
+
+  private def subscribe(): Unit = {
+    logging.info(this, s"subscribing to mesos master at ${mesosMaster}")
+    //subscribe mesos actor to mesos event stream
+    //TODO: subscribe failure should make invoker "unhealthy"
+    mesosClientActor
+      .ask(Subscribe)(subscribeTimeout)
+      .mapTo[SubscribeComplete]
+      .onComplete({
+        case Success(complete) =>
+          isSubscribed = true
+          logging.info(this, s"subscribe completed successfully...${complete}")
+        case Failure(e) => logging.error(this, s"subscribe failed...${e}")
+      })
+  }
+
+  override def createContainer(tid: TransactionId,
+                               name: String,
+                               actionImage: ExecManifest.ImageName,
+                               userProvidedImage: Boolean,
+                               memory: ByteSize)(implicit config: WhiskConfig, logging: Logging): Future[Container] = {
+    implicit val transid = tid
+    val image = if (userProvidedImage) {
+      actionImage.publicImageName
+    } else {
+      actionImage.localImageName(config.dockerRegistry, config.dockerImagePrefix, Some(config.dockerImageTag))
+    }
+
+    logging.info(this, s"using Mesos to create a container with image ${image}...")
+    MesosTask.create(
+      mesosClientActor,
+      mesosConfig,
+      taskIdGenerator,
+      tid,
+      image = image,
+      userProvidedImage = userProvidedImage,
+      memory = memory,
+      cpuShares = config.invokerCoreShare.toInt,
+      environment = Map("__OW_API_HOST" -> config.wskApiHost),
+      network = config.invokerContainerNetwork,
+      dnsServers = config.invokerContainerDns,
+      name = Some(name),
+      parameters)
+  }
+
+  override def init(): Unit = Unit
+
+  /** cleanup any remaining Containers; should block until complete; should ONLY be run at shutdown */
+  override def cleanup(): Unit = {
+    val complete: Future[Any] = mesosClientActor.ask(Teardown)(20.seconds)
+    Await.result(complete, 25.seconds)
```

Comment:
if the await times out, this method will throw an exception. is that intended?

Reply:
I'll change this to log an error that teardown failed